src/main/java/org/jenkinsci/plugins/junitrealtimetestreporter/PipelineAttacher.java

```diff
@@ -0,0 +1,211 @@
+import org.jenkinsci.plugins.workflow.flow.GraphListener;
+import org.jenkinsci.plugins.workflow.graph.BlockEndNode;
+import org.jenkinsci.plugins.workflow.graph.BlockStartNode;
+import org.jenkinsci.plugins.workflow.graph.FlowNode;
+import org.jenkinsci.plugins.workflow.graph.StepNode;
+import org.jenkinsci.plugins.workflow.graphanalysis.DepthFirstScanner;
+
+/**
+ * Attaches distinct test result actions to each {@code node} (or {@code ws}) block in a job with this plugin activated.
+ * Looks for all {@code junit} steps in the {@code lastSuccessfulBuild} so it can guess at test patterns to archive.
+ */
+@Extension
+public class PipelineAttacher implements GraphListener {
+
+    private static final Logger LOGGER = Logger.getLogger(PipelineAttacher.class.getName());
+
+    @Override
+    public void onNewHead(FlowNode node) {
+        if (LOGGER.isLoggable(Level.FINER)) {
+            LOGGER.log(Level.FINER, "notified of {0} ~ {1}", new Object[] {node, node.getDisplayFunctionName()});
+        }
+        if (node instanceof BlockStartNode) {
+            for (FlowNode node2 : node.getParents()) { // do not pay attention until the body is invoked, when WorkspaceAction is attached
+                LOGGER.log(Level.FINER, "considering of {0} ~ {1}", new Object[] {node2, node2.getDisplayFunctionName()});
+                WorkspaceAction wsa = node2.getPersistentAction(WorkspaceAction.class);
+                if (wsa != null) {
+                    FilePath workspace = wsa.getWorkspace();
+                    if (workspace == null) {
+                        LOGGER.fine("no workspace");
+                        return;
+                    }
+                    Queue.Executable executable;
+                    try {
+                        executable = node2.getExecution().getOwner().getExecutable();
+                    } catch (IOException x) {
+                        LOGGER.log(Level.WARNING, null, x);
+                        return;
+                    }
+                    if (executable instanceof Run) {
+                        Run<?, ?> run = (Run<?, ?>) executable;
+                        Job<?, ?> job = run.getParent();
+                        if (!PerJobConfiguration.isActive(job)) {
+                            LOGGER.log(Level.FINE, "inactive in {0}", job);
+                            return;
+                        }
+                        LOGGER.log(Level.FINE, "encountered {0} in {1} entering {2}", new Object[] {workspace, run, node2.getDisplayFunctionName()});
+                        Run<?, ?> last = job.getLastSuccessfulBuild();
+                        if (!(last instanceof FlowExecutionOwner.Executable)) {
+                            LOGGER.log(Level.FINE, "no lastSuccessfulBuild in {0}", job);
+                            return;
+                        }
+                        FlowExecutionOwner owner = ((FlowExecutionOwner.Executable) last).asFlowExecutionOwner();
```

Comment:
Would be nice to clarify that `owner` and `exec` here are from the last successful build in their naming - I glossed over that initially and got nervous about the `DepthFirstScanner` invocation below, since I know doing scanning against running builds is questionable. But since this is against a complete build, that isn't a problem.